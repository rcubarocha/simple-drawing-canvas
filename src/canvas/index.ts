import { RGBColor } from '../color';

interface ICoords {
    x: number,
    y: number,
}

type TouchEventsMap = Pick<HTMLElementEventMap, 'touchstart' | 'touchmove' | 'touchend' | 'touchcancel'>;

export type CanvasToolType = 'pen' | 'bucket' | 'eraser' | 'clear';

type CanvasAction = {
    tool: CanvasToolType,
    start?: ICoords,
    end?: ICoords,
    size?: number,
    color?: string,
}

interface IDrawingCanvas {
    canvasElement: HTMLCanvasElement,
    width: number,
    height: number,
    scale: number,
    toolType: CanvasToolType
    brushSize: number,
    brushColor: RGBColor,
    lastCoords: ICoords,
    toolDown: boolean,
    actionsHistory: CanvasAction[],
    historyIndex: number,
    initCanvas(): void,
}

type MouseEventCallback = (e: MouseEvent) => void;

export class DrawingCanvasController implements IDrawingCanvas {
    canvasElement: HTMLCanvasElement;

    width: number;

    height: number;

    scale: number = 1;

    toolType: CanvasToolType = 'pen';

    brushSize: number = 5;

    brushColor: RGBColor = new RGBColor(0, 0, 0);

    lastCoords: ICoords = {
      x: 0,
      y: 0,
    };

    toolDown: boolean = false;

    actionsHistory: CanvasAction[] = [];

    historyIndex: number = 0;

    // boundMouseCallback = this.onCanvasEvent.bind(this);
    // boundTouchCallback = this.onTouchEvent.bind(this);

    canvasFocusCallback = (e: TouchEvent) => {
      if (e.target === this.canvasElement) {
        e.preventDefault();
      }
    }

    static eventMap = new Map<keyof TouchEventsMap, keyof HTMLElementEventMap>([
      ['touchstart', 'mousedown'],
      ['touchmove', 'mousemove'],
      ['touchend', 'mouseup'],
      ['touchcancel', 'mouseup'],
    ]);

    toolFuncs = new Map<CanvasToolType, MouseEventCallback>([
      ['pen', this.penTool.bind(this)],
      ['clear', this.clearTool.bind(this)],
      ['bucket', this.bucketTool.bind(this)],
      ['eraser', this.eraserTool.bind(this)],
    ]);

    constructor(canvasElement: HTMLCanvasElement, width: number, height: number) {
      this.canvasElement = canvasElement;
      this.width = width;
      this.height = height;

      this.initCanvas();
    }

    // eslint-disable-next-line class-methods-use-this
    drawLine(ctx: CanvasRenderingContext2D, from: ICoords, to: ICoords, width: number, color: string) {
      ctx.beginPath();
      ctx.lineCap = 'round';
      ctx.lineJoin = 'round';
      ctx.lineWidth = width;
      ctx.moveTo(from.x, from.y);
      ctx.lineTo(to.x, to.y);
      ctx.strokeStyle = color;
      ctx.stroke();
      ctx.closePath();
    }

    performCanvasAction(action: CanvasAction) {
      const ctx = this.canvasElement.getContext('2d')!;

      switch (action.tool) {
        case 'pen':
          ctx.globalCompositeOperation = 'source-over';

          this.drawLine(ctx, action.start!, action.end!, action.size!, action.color!);
          break;
        case 'clear':

          ctx.globalCompositeOperation = 'source-over';

          ctx.clearRect(0, 0, this.canvasElement.width, this.canvasElement.height);
          break;
        case 'bucket':
          ctx.globalCompositeOperation = 'source-over';

          ctx.fillStyle = action.color!;
          ctx.fillRect(0, 0, this.canvasElement.width, this.canvasElement.height);
          break;
        case 'eraser':
          ctx.globalCompositeOperation = 'destination-out';

          this.drawLine(ctx, action.start!, action.end!, action.size!, '#ffffff');
          break;
        default:
          break;
      }
    }

    penTool(e: MouseEvent) {
      if (e.type === 'mousedown') {
        this.toolDown = true;
        const { x, y } = this.canvasElement.getBoundingClientRect();
        this.lastCoords.x = (e.clientX - x) * this.scale;
        this.lastCoords.y = (e.clientY - y) * this.scale;
      } else if (e.type === 'mousemove') {
        if (!this.toolDown) return;

        const { x, y } = this.canvasElement.getBoundingClientRect();

        const newCoords: ICoords = {
          x: (e.clientX - x) * this.scale,
          y: (e.clientY - y) * this.scale,
        };

        const action: CanvasAction = {
          tool: 'pen',
          start: { ...this.lastCoords },
          end: { ...newCoords },
          size: this.brushSize,
          color: this.brushColor.toHex(),
        };

        this.actionsHistory.push(action);

        this.performCanvasAction(action);

        this.lastCoords = newCoords;
      } else if (e.type === 'mouseup') {
        if (!this.toolDown) return;

        if (e.target === this.canvasElement) {
          const { x, y } = this.canvasElement.getBoundingClientRect();

          const newCoords: ICoords = {
            x: (e.clientX - x) * this.scale,
            y: (e.clientY - y) * this.scale,
          };

          const action: CanvasAction = {
            tool: 'pen',
            start: { ...this.lastCoords },
            end: { ...newCoords },
            size: this.brushSize,
            color: this.brushColor.toHex(),
          };

          this.actionsHistory.push(action);

          this.performCanvasAction(action);

          this.lastCoords = newCoords;
        }

        this.toolDown = false;
      }
    }

    clearTool(e: MouseEvent) {
      if (e.type === 'mousedown') {
        this.toolDown = true;
      } else if (e.type === 'mouseup') {
        if (!this.toolDown) return;

        // Only clear the canvas if the click release happens inside the canvas
        if (e.target === this.canvasElement) {
          const action: CanvasAction = { tool: 'clear' };

          this.actionsHistory.push(action);

          this.performCanvasAction(action);
        }

        this.toolDown = false;
      }
    }

    bucketTool(e: MouseEvent) {
      if (e.type === 'mousedown') {
        this.toolDown = true;
      } else if (e.type === 'mouseup') {
        if (!this.toolDown) return;

        // Only clear the canvas if the click release happens inside the canvas
        if (e.target === this.canvasElement) {
          // const action: CanvasAction = { tool: 'bucket', color: this.brushColor.toHex() };

          // this.actionsHistory.push(action);

          // this.performCanvasAction(action);

          // this.playDrawing();
          this.performAllCanvasActions();
        }

        this.toolDown = false;
      }
    }

    eraserTool(e: MouseEvent) {
      if (e.type === 'mousedown') {
        this.toolDown = true;
        const { x, y } = this.canvasElement.getBoundingClientRect();
        this.lastCoords.x = (e.clientX - x) * this.scale;
        this.lastCoords.y = (e.clientY - y) * this.scale;
      } else if (e.type === 'mousemove') {
        if (!this.toolDown) return;

        const { x, y } = this.canvasElement.getBoundingClientRect();

        const newCoords = {
          x: (e.clientX - x) * this.scale,
          y: (e.clientY - y) * this.scale,
        };

        const action: CanvasAction = {
          tool: 'eraser', start: { ...this.lastCoords }, end: { ...newCoords }, size: this.brushSize,
        };

        this.actionsHistory.push(action);

        this.performCanvasAction(action);

        this.lastCoords = newCoords;
      } else if (e.type === 'mouseup') {
        if (!this.toolDown) return;

        if (e.target === this.canvasElement) {
          const { x, y } = this.canvasElement.getBoundingClientRect();

          const newCoords = {
            x: (e.clientX - x) * this.scale,
            y: (e.clientY - y) * this.scale,
          };

          const action: CanvasAction = {
            tool: 'eraser', start: { ...this.lastCoords }, end: { ...newCoords }, size: this.brushSize,
          };

          this.actionsHistory.push(action);

          this.performCanvasAction(action);

          this.lastCoords = newCoords;
        }

        this.toolDown = false;
      }
    }

    performAllCanvasActions() {
      const action: CanvasAction = { tool: 'clear' };

      this.performCanvasAction(action);

      this.actionsHistory.pop();

      this.actionsHistory.forEach((ca) => {
        this.performCanvasAction(ca);
      });
    }

    playDrawing() {
      const action: CanvasAction = { tool: 'clear' };

      this.performCanvasAction(action);

      this.historyIndex = 0;

      const interval = setInterval(() => {
        if (this.actionsHistory.length > 0) this.performCanvasAction(this.actionsHistory[this.historyIndex]!);

        if (this.historyIndex < this.actionsHistory.length - 1) {
          this.historyIndex += 1;
        } else {
          clearInterval(interval);
        }
      }, 5);
    }

    onCanvasEvent = (e: MouseEvent) => {
        // console.log('-- On Canvas Event --');
        // console.log({ type: e.type, target: e.target, current: e.currentTarget });
        this.toolFuncs.get(this.toolType)!(e);
    }

    onTouchEvent = (e: TouchEvent) => {
      // console.log('-- On Touch Event --');
      // console.log({ type: e.type, target: e.target, current: e.currentTarget });

      const me = new MouseEvent(DrawingCanvasController.eventMap.get(e.type as keyof TouchEventsMap)!, {
        clientX: e.changedTouches[0].clientX,
        clientY: e.changedTouches[0].clientY,
        bubbles: true,
      });

      /**
       * TouchEnd always targets the target of the preceding TouchStart event
       * even if the pointer has moved away from that element
       * We get the element over which the TouchEnd happened using elementFromPoint()
       * so we can accurately trigger the synthetic mouse event
       */
      const doc = this.canvasElement.ownerDocument!;
      const touchEndElement = doc.elementFromPoint(me.clientX, me.clientY)!;

      if (doc && touchEndElement) {
        touchEndElement.dispatchEvent(me);
      }

      // if (e.type === 'touchend' || e.type === 'touchcancel') {
      //     this.canvasElement.ownerDocument!.dispatchEvent(me);
      //     console.log('Dispathcing MouseUP from TouchEnd/Cancel')
      //     // this.canvasElement.dispatchEvent(me);
      // }
      // else {
      //     this.canvasElement.dispatchEvent(me);
      // }
    }

    onWindowResize = () => {
      this.scale = this.width / this.canvasElement.offsetWidth;

      this.canvasElement.style.height = `${this.height / this.scale}px`;
    }

    reset() {
      this.actionsHistory = [];
      this.historyIndex = 0;
      this.lastCoords = {
        x: 0,
        y: 0,
      };
      this.toolDown = false;
    }

    teardown() {
      DrawingCanvasController.eventMap.forEach((_, k) => {
        if (k === 'touchend' || k === 'touchcancel') {
                this.canvasElement.ownerDocument!.removeEventListener(k, this.onTouchEvent);
        } else {
          this.canvasElement.removeEventListener(k, this.onTouchEvent);
        }

            // Prevent scrolling on touch interfaces when event occurs on canvas
            this.canvasElement.ownerDocument!.body.removeEventListener(k, this.canvasFocusCallback, { capture: false });
      });

      const w = this.canvasElement.ownerDocument!.defaultView;

      if (w) {
        w.removeEventListener('resize', this.onWindowResize);
      }

      this.canvasElement.removeEventListener('mousedown', this.onCanvasEvent);
      this.canvasElement.removeEventListener('mousemove', this.onCanvasEvent);
      this.canvasElement.ownerDocument!.removeEventListener('mouseup', this.onCanvasEvent);
    }

    initCanvas() {
      this.canvasElement.width = this.width;
      this.canvasElement.height = this.height;

      const w = this.canvasElement.ownerDocument!.defaultView;

      // w && (w.onresize = this.onWindowResize);
      if (w) {
        w.addEventListener('resize', this.onWindowResize);
      }

      this.onWindowResize();

      DrawingCanvasController.eventMap.forEach((_, k) => {
        if (k === 'touchend' || k === 'touchcancel') {
                this.canvasElement.ownerDocument!.addEventListener(k, this.onTouchEvent);
        } else {
          this.canvasElement.addEventListener(k, this.onTouchEvent);
        }

        // Prevent scrolling on touch interfaces when event occurs on canvas
        const { body } = this.canvasElement.ownerDocument!;

        body.addEventListener(k, this.canvasFocusCallback, { capture: false, passive: false });
      });

      this.canvasElement.addEventListener('mousedown', this.onCanvasEvent);
      this.canvasElement.addEventListener('mousemove', this.onCanvasEvent);
      // Setting listener to the document so lifting the mouse button anywhere will register as the tool being lifted
      this.canvasElement.ownerDocument!.addEventListener('mouseup', this.onCanvasEvent);
    }
}
