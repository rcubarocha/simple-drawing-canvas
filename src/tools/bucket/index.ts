import type {
  ToolMouseEventCallback, ToolActionStepCallback, StrokeFillStyle,
} from '../../canvas';
import { getCanvasCoordsFromEvent } from '../../utils';

export interface BucketTool {
  style: StrokeFillStyle
}

export const bucketMouseEventCallback: ToolMouseEventCallback<BucketTool> = function bucketMouseEventCallback(
  event, canvas, canvasConfig, toolConfig, actionHistory,
) {
  if (event.type === 'mousedown') {
    const state = 'down';

    const eCoords = getCanvasCoordsFromEvent(event, canvas, canvasConfig);

    return {
      endCurrentAction: false,
      actionStep: {
        tool: toolConfig,
        coords: eCoords,
        state,
      },
      replacePrevStep: false,
    };
  }

  if (event.type === 'mouseup') {
    if (actionHistory.steps.length < 1) {
      return null;
    }

    const prevToolState = actionHistory.steps[actionHistory.steps.length - 1].state;

    if (prevToolState !== 'down') {
      throw Error('Inconsistent State');
    }

    // If the mouse up event happens outside the canvas, invalidate the whole action
    if (event.target !== canvas) {
      throw Error('Up Event Outside Canvas');
    } else {
      const state = 'up';

      const eCoords = getCanvasCoordsFromEvent(event, canvas, canvasConfig);

      return {
        endCurrentAction: true,
        actionStep: {
          tool: toolConfig,
          coords: eCoords,
          state,
        },
        replacePrevStep: false,
      };
    }
  }

  return null;
};

export const bucketDrawingCallback: ToolActionStepCallback<BucketTool> = function bucketDrawingCallback(
  actionStep, _, canvas,
) {
  if (actionStep.state === 'down') {
    // If the action is from the mousedown event, do nothing.
    return;
  }

  if (actionStep.state === 'up') {
    const ctx = canvas.getContext('2d')!;

    ctx.globalCompositeOperation = 'source-over';

    ctx.fillStyle = actionStep.tool.style;
    ctx.fillRect(0, 0, canvas.width, canvas.height);

    return;
  }

  throw Error(`Unrecognized Bucket Tool State: ${actionStep.state}`);
};
