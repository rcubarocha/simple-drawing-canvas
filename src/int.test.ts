import { PenTool, penMouseEventCallback, penDrawingCallback } from "./tools/pen";
import { DrawingCanvasController } from "./canvas";

describe('tool & controller integration', () => {
  let canvas: HTMLCanvasElement;
  let penConfig: PenTool;
  let controller: DrawingCanvasController<'pen', { pen: PenTool }>;

  beforeEach(() => {
    canvas = document.createElement('canvas');

    penConfig = {
      size: 15,
      style: '#2495A4',
    }

    // jsdom does not implement layout-related properties
    jest.spyOn(canvas, 'offsetWidth', 'get').mockReturnValue(100);
    
    canvas.getBoundingClientRect = jest.fn(() => ({
      x: 50,
      y: 50,
      width: 100,
      height: 100,
      top: 50,
      right: 150,
      bottom: 150,
      left: 50,
    } as DOMRect));

    controller = new DrawingCanvasController(canvas, 300, 300, 'pen');

    controller.addTool('pen', penMouseEventCallback, penDrawingCallback, penConfig);
  });

  test('draw on canvas through mouse events', () => {
    const mouseEvents: MouseEvent[] = [
      new MouseEvent('mousedown', {
        clientX: 70,
        clientY: 130,
      }),
      new MouseEvent('mousemove', {
        clientX: 100,
        clientY: 70,
      }),
      new MouseEvent('mousemove', {
        clientX: 130,
        clientY: 130,
      }),
      new MouseEvent('mousemove', {
        clientX: 70,
        clientY: 130,
      }),
      new MouseEvent('mouseup', {
        clientX: 70,
        clientY: 130,
      }),
    ];

    mouseEvents.forEach((ev) => {
      canvas.dispatchEvent(ev);
    });

    expect(canvas.toDataURL()).toMatchSnapshot();
  });

  test('draw on canvas through touch events', () => {
    // jsdom does not implement document.elementFromPoint
    // Provide our own mock implementation
    document['elementFromPoint'] = (x, y) => { return canvas }

    // jsdom does not define Touch constructor so we assert objects with necessary info
    const touchEvents: TouchEvent[] = [
      new TouchEvent('touchstart', {
        changedTouches: [
          {
            clientX: 70,
            clientY: 130,
          } as Touch
        ]
      }),
      new TouchEvent('touchmove', {
        changedTouches: [
          {
            clientX: 100,
            clientY: 70,
          } as Touch
        ]
      }),
      new TouchEvent('touchmove', {
        changedTouches: [
          {
            clientX: 130,
            clientY: 130,
          } as Touch
        ]
      }),
      new TouchEvent('touchmove', {
        changedTouches: [
          {
            clientX: 70,
            clientY: 130,
          } as Touch
        ]
      }),
      new TouchEvent('touchend', {
        changedTouches: [
          {
            clientX: 70,
            clientY: 130,
          } as Touch
        ]
      }),
    ];

    touchEvents.forEach((ev) => {
      canvas.dispatchEvent(ev);
    });

    expect(canvas.toDataURL()).toMatchSnapshot();
  });
});