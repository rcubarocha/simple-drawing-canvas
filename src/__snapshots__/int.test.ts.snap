// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tool & controller integration draw on canvas through mouse events 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAYAAAB5fY51AAAABmJLR0QA/wD/AP+gvaeTAAAKK0lEQVR4nO3bO4+cVx3H8XNm1pIRDW+ABttrO6FOwatw4YaLFaUISCFRSiRAKSgoUiAlIZEgRRSciMZF8ioQPYq9DqGhQaKhQQmyZw5F9ji+7GUuz/Oc85zz+XQreUf/lWZ+/s6OHQIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA8KZY+gHZcf+uDKw8vXHg5hXg9hBBiSHcPHjx4/+7rL31e+jbaYLDY2+E7d66l5VefpJSuPvuUSiHGeBRXF2/cf+3mvSIH0gyDxV4u/eH2jbiKd0IKyzP/YAyrtEw3v/jZrU8mOo0GGSx2dvjOnWvr5Vd/O3esshhWi9XF7ystdrUofQDztV7879ONxyqEEFJYrhZffjriSTTOYLGTw7c/fCGE9eG23xdDPPz6e2F7BoudpOXizV1/o7A+WLw58Dl0wu+w2Nqld29fjiEehd3/wlsvY3ru6JVbR0PeRfsUFluLMb4R9nvuLFZp8cuh7qEfCoutfO/3Hx4u4vKzkMJyz2fPap1Wz/3j1RfvD3QaHVBYbGWxWP46hDxWaZ+HWh4/FmxMYbGxR3UVtvinDGdTWWxFYbGxR3U1nOUiHvxqwMejcQqLjRx/Mng3hHAw8EOvljE97xNDNqGw2MjxJ4NDj1UIISx9YsimFBbnGrGuMpXFRhQW5xqxrjKVxUYUFmeaoK4ylcW5FBZnmqCuMpXFuRQWp5qwrjKVxZkUFqeasK4ylcWZFBYnKlBXmcriVAqLExWoq0xlcSqFxTMK1lWmsjiRwuIZA9TVKoW02uP7VRYnUlg8YZi6in8KMcWQwq09TlFZPENh8YQh6moZ179NKf0mhPBwj8dRWTzDYPHIpXdvXw4p/HC/R4kfH71y6+iLn9/6e4jhz/s9Vvrx1fduX93vMWiJweKRoeoqf6GyGJrBIoQwbF3lr1QWQzNYhBCGr6tMZTEkg8UodZWpLIZksBitrjKVxVAMVufGrKtMZTEUg9W5sesqU1kMwWB1bIq6ylQWQzBYHZuqrjKVxb4MVqemrKtMZbEvg9WpqesqU1nsw2B1qERdZSqLfRisDpWqq0xlsSuD1ZmSdZWpLHZlsDpTuq4ylcUuDFZHaqirTGWxC4PVkVrqKlNZbMtgdaKmuspUFtsyWJ2ora4ylcU2DFYHaqyrTGWxDYPVgVrrKlNZbMpgNa7muspUFpsyWI2rva4ylcUmDFbD5lBXmcpiEwarYXOpq0xlcR6D1ag51VWmsjiPwWrU3OoqU1mcxWA1aI51lakszmKwGjTXuspUFqcxWI2Zc11lKovTGKzGzL2uMpXFSQxWQ1qoq0xlcRKD1ZBW6ipTWTzNYDWipbrKVBZPM1iNaK2uMpXF4wxWA1qsq0xl8TiD1YBW6ypTWWQGa+ZarqtMZZEZrJlrva4ylUUIBmvWeqirTGURgsGatV7qKlNZGKyZ6qmuMpWFwZqp3uoqU1l9M1gz1GNdZSqrbwZrhnqtq0xl9ctgzUzPdZWprH4ZrJnpva4yldUngzUj6uobKqtPBmtG1NWTVFZ/DNZMqKtnqaz+GKyZUFcnU1l9MVgzoK5Op7L6YrBmQF2dTWX1w2BVTl2dT2X1w2BVTl1tRmX1wWBVTF1tTmX1wWBVTF1tR2W1z2BVSl1tT2W1z2BVSl3tRmW1zWBVSF3tTmW1zWBVSF3tR2W1y2BVRl3tT2W1y2BVRl0NQ2W1yWBVRF0NR2W1yWBVRF0NS2W1x2BVQl0NT2W1x2BVQl2NQ2W1xWBVQF2NR2W1xWBVQF2NS2W1w2AVpq7Gp7LaYbAKU1fTUFltMFgFqavpqKw2GKyC1NW0VNb8GaxC1NX0VNb8GaxC1FUZKmveDFYB6qoclTVvBqsAdVWWypovgzUxdVWeypovgzUxdVUHlTVPBmtC6qoeKmueDNaE1FVdVNb8GKyJqKv6qKz5MVgTUVd1UlnzYrAmoK7qpbLmxWBNQF3VTWXNh8Eambqqn8qaD4M1MnU1DyprHgzWiNTVfKiseTBYI1JX86Ky6mewRqKu5kdl1c9gjURdzZPKqpvBGoG6mi+VVTeDNQJ1NW8qq14Ga2Dqav5UVr0M1sDUVRtUVp0M1oDUVTtUVp0M1oDUVVtUVn0M1kDUVXtUVn0M1kDUVZtUVl0M1gDUVbtUVl0M1gDUVdtUVj0M1p7UVftUVj0M1p7UVR9UVh0M1h7UVT9UVh0M1h7UVV9UVnkGa0fqqj8qqzyDtSN11SeVVZbB2oG66pfKKstg7UBd9U1llWOwtqSuUFnlGKwtqStCUFmlGKwtqCsylVWGwdqCuuJxKmt6BmtD6oqnqazpGawNqStOorKmZbA2oK44jcqalsHagLriLCprOgbrHOqK86is6Risc6grNqGypmGwzqCu2JTKmobBOoO6Yhsqa3wG6xTqim2prPEZrFOoK3ahssZlsE6grtiVyhqXwTqBumIfKms8Busp6op9qazxGKynqCuGoLLGYbAeo64Yisoah8F6jLpiSCpreAbrmLpiaCpreAbrmLpiDCprWAYrqCvGo7KGZbCCumJcKms43Q+WumJsKms43Q+WumIKKmsYXQ+WumIqKmsYXQ+WumJKKmt/3Q6WumJqKmt/3Q6WuqIElbWfLgdLXVGKytpPl4OlrihJZe2uu8FSV5SmsnbX3WCpK2qgsnbT1WCpK2qhsnbT1WCpK2qisrbXzWCpK2qjsrbXzWCpK2qksrbTxWCpK2qlsrbTxWCpK2qmsjbX/GCpK2qnsjbX/GCpK+ZAZW2m6cFSV8yFytrMPuVRvRjjGyHt9TOuQ0j/vvzex78Y7Cg4RQzpXymEddg9JHJlvTjgWVWJpQ8Yy6V3b1+OYXE3hHQQQgoN/6gQvnmOx9Uyrp9v9V1Bs28JYwy/+3qsQjBWtC8/x9NyFeKbRU8ZUZOv5MO3P35hvUx/LX0HlHKQFj+49+qP/lL6jqE1WVjrZfqo9A1Q0sO4/rD0DWNobrCuv/XBlRDSldJ3QFnpytevhbY0N1gPFwc/bfSdLmwhHr8W2tLcYKXl4vnSN0AVFovnSp8wtOYGK6aQSt8AdUjNvRaaG6ywXn9W+gSowjrdLX3C0JobrIP1wz8GkUX30vFroS1N/nb6ynsf3UspNP1/quAsMYajz1/5ybXSdwytucIKIYS4ungjxLAqfQcUEcMqri7eKH3GGJocrPuv3byXlumm0aI7MazSMt28/9rNe6VPGUOTbwmzw3fuXEvLrz5JKV1t/EeleynEGI/i6uKNVscqhE5exdff+uDKwwsXXk4hXk8pfXuRwn/WIfwzhvRl6dtgWynEby1C+O46hu/EGP8bQ7p78ODB+3dff+nz0rcBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADA1P4PSDLCo1HUi/sAAAAASUVORK5CYII="`;

exports[`tool & controller integration draw on canvas through touch events 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAYAAAB5fY51AAAABmJLR0QA/wD/AP+gvaeTAAAKK0lEQVR4nO3bO4+cVx3H8XNm1pIRDW+ABttrO6FOwatw4YaLFaUISCFRSiRAKSgoUiAlIZEgRRSciMZF8ioQPYq9DqGhQaKhQQmyZw5F9ji+7GUuz/Oc85zz+XQreUf/lWZ+/s6OHQIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA8KZY+gHZcf+uDKw8vXHg5hXg9hBBiSHcPHjx4/+7rL31e+jbaYLDY2+E7d66l5VefpJSuPvuUSiHGeBRXF2/cf+3mvSIH0gyDxV4u/eH2jbiKd0IKyzP/YAyrtEw3v/jZrU8mOo0GGSx2dvjOnWvr5Vd/O3esshhWi9XF7ystdrUofQDztV7879ONxyqEEFJYrhZffjriSTTOYLGTw7c/fCGE9eG23xdDPPz6e2F7BoudpOXizV1/o7A+WLw58Dl0wu+w2Nqld29fjiEehd3/wlsvY3ru6JVbR0PeRfsUFluLMb4R9nvuLFZp8cuh7qEfCoutfO/3Hx4u4vKzkMJyz2fPap1Wz/3j1RfvD3QaHVBYbGWxWP46hDxWaZ+HWh4/FmxMYbGxR3UVtvinDGdTWWxFYbGxR3U1nOUiHvxqwMejcQqLjRx/Mng3hHAw8EOvljE97xNDNqGw2MjxJ4NDj1UIISx9YsimFBbnGrGuMpXFRhQW5xqxrjKVxUYUFmeaoK4ylcW5FBZnmqCuMpXFuRQWp5qwrjKVxZkUFqeasK4ylcWZFBYnKlBXmcriVAqLExWoq0xlcSqFxTMK1lWmsjiRwuIZA9TVKoW02uP7VRYnUlg8YZi6in8KMcWQwq09TlFZPENh8YQh6moZ179NKf0mhPBwj8dRWTzDYPHIpXdvXw4p/HC/R4kfH71y6+iLn9/6e4jhz/s9Vvrx1fduX93vMWiJweKRoeoqf6GyGJrBIoQwbF3lr1QWQzNYhBCGr6tMZTEkg8UodZWpLIZksBitrjKVxVAMVufGrKtMZTEUg9W5sesqU1kMwWB1bIq6ylQWQzBYHZuqrjKVxb4MVqemrKtMZbEvg9WpqesqU1nsw2B1qERdZSqLfRisDpWqq0xlsSuD1ZmSdZWpLHZlsDpTuq4ylcUuDFZHaqirTGWxC4PVkVrqKlNZbMtgdaKmuspUFtsyWJ2ora4ylcU2DFYHaqyrTGWxDYPVgVrrKlNZbMpgNa7muspUFpsyWI2rva4ylcUmDFbD5lBXmcpiEwarYXOpq0xlcR6D1ag51VWmsjiPwWrU3OoqU1mcxWA1aI51lakszmKwGjTXuspUFqcxWI2Zc11lKovTGKzGzL2uMpXFSQxWQ1qoq0xlcRKD1ZBW6ipTWTzNYDWipbrKVBZPM1iNaK2uMpXF4wxWA1qsq0xl8TiD1YBW6ypTWWQGa+ZarqtMZZEZrJlrva4ylUUIBmvWeqirTGURgsGatV7qKlNZGKyZ6qmuMpWFwZqp3uoqU1l9M1gz1GNdZSqrbwZrhnqtq0xl9ctgzUzPdZWprH4ZrJnpva4yldUngzUj6uobKqtPBmtG1NWTVFZ/DNZMqKtnqaz+GKyZUFcnU1l9MVgzoK5Op7L6YrBmQF2dTWX1w2BVTl2dT2X1w2BVTl1tRmX1wWBVTF1tTmX1wWBVTF1tR2W1z2BVSl1tT2W1z2BVSl3tRmW1zWBVSF3tTmW1zWBVSF3tR2W1y2BVRl3tT2W1y2BVRl0NQ2W1yWBVRF0NR2W1yWBVRF0NS2W1x2BVQl0NT2W1x2BVQl2NQ2W1xWBVQF2NR2W1xWBVQF2NS2W1w2AVpq7Gp7LaYbAKU1fTUFltMFgFqavpqKw2GKyC1NW0VNb8GaxC1NX0VNb8GaxC1FUZKmveDFYB6qoclTVvBqsAdVWWypovgzUxdVWeypovgzUxdVUHlTVPBmtC6qoeKmueDNaE1FVdVNb8GKyJqKv6qKz5MVgTUVd1UlnzYrAmoK7qpbLmxWBNQF3VTWXNh8Eambqqn8qaD4M1MnU1DyprHgzWiNTVfKiseTBYI1JX86Ky6mewRqKu5kdl1c9gjURdzZPKqpvBGoG6mi+VVTeDNQJ1NW8qq14Ga2Dqav5UVr0M1sDUVRtUVp0M1oDUVTtUVp0M1oDUVVtUVn0M1kDUVXtUVn0M1kDUVZtUVl0M1gDUVbtUVl0M1gDUVdtUVj0M1p7UVftUVj0M1p7UVR9UVh0M1h7UVT9UVh0M1h7UVV9UVnkGa0fqqj8qqzyDtSN11SeVVZbB2oG66pfKKstg7UBd9U1llWOwtqSuUFnlGKwtqStCUFmlGKwtqCsylVWGwdqCuuJxKmt6BmtD6oqnqazpGawNqStOorKmZbA2oK44jcqalsHagLriLCprOgbrHOqK86is6Risc6grNqGypmGwzqCu2JTKmobBOoO6Yhsqa3wG6xTqim2prPEZrFOoK3ahssZlsE6grtiVyhqXwTqBumIfKms8Busp6op9qazxGKynqCuGoLLGYbAeo64Yisoah8F6jLpiSCpreAbrmLpiaCpreAbrmLpiDCprWAYrqCvGo7KGZbCCumJcKms43Q+WumJsKms43Q+WumIKKmsYXQ+WumIqKmsYXQ+WumJKKmt/3Q6WumJqKmt/3Q6WuqIElbWfLgdLXVGKytpPl4OlrihJZe2uu8FSV5SmsnbX3WCpK2qgsnbT1WCpK2qhsnbT1WCpK2qisrbXzWCpK2qjsrbXzWCpK2qksrbTxWCpK2qlsrbTxWCpK2qmsjbX/GCpK2qnsjbX/GCpK+ZAZW2m6cFSV8yFytrMPuVRvRjjGyHt9TOuQ0j/vvzex78Y7Cg4RQzpXymEddg9JHJlvTjgWVWJpQ8Yy6V3b1+OYXE3hHQQQgoN/6gQvnmOx9Uyrp9v9V1Bs28JYwy/+3qsQjBWtC8/x9NyFeKbRU8ZUZOv5MO3P35hvUx/LX0HlHKQFj+49+qP/lL6jqE1WVjrZfqo9A1Q0sO4/rD0DWNobrCuv/XBlRDSldJ3QFnpytevhbY0N1gPFwc/bfSdLmwhHr8W2tLcYKXl4vnSN0AVFovnSp8wtOYGK6aQSt8AdUjNvRaaG6ywXn9W+gSowjrdLX3C0JobrIP1wz8GkUX30vFroS1N/nb6ynsf3UspNP1/quAsMYajz1/5ybXSdwytucIKIYS4ungjxLAqfQcUEcMqri7eKH3GGJocrPuv3byXlumm0aI7MazSMt28/9rNe6VPGUOTbwmzw3fuXEvLrz5JKV1t/EeleynEGI/i6uKNVscqhE5exdff+uDKwwsXXk4hXk8pfXuRwn/WIfwzhvRl6dtgWynEby1C+O46hu/EGP8bQ7p78ODB+3dff+nz0rcBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADA1P4PSDLCo1HUi/sAAAAASUVORK5CYII="`;
